labelbox/data/annotation_types/collection.py

```diff
@@ -0,0 +1,273 @@
+        if self._index == len(self._data):
+            self._index = 0
+            raise StopIteration
+
+        value = self._data[self._index]
+        self._index += 1
+        return value
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+    def __getitem__(self, idx: int) -> Label:
+        return self._data[idx]
+
+    def _apply_threaded(self, fns, max_concurrency, *args):
+        futures = []
+        with ThreadPoolExecutor(max_workers=max_concurrency) as executor:
+            for fn in fns:
+                futures.append(executor.submit(fn, *args))
+            for future in tqdm(as_completed(futures)):
+                yield future.result()
+
+
+class LabelGenerator(PrefetchGenerator):
+    """
+    A container for interacting with a collection of labels.
+
+    Use this class if you have larger data. It is slightly harder to work with
+    than the LabelCollection but will be much more memory efficient.
+    """
+
+    def __init__(self, data: Generator[Label, None, None], *args, **kwargs):
+        self._fns = {}
+        super().__init__(data, *args, **kwargs)
+
+    def as_collection(self) -> "LabelCollection":
+        return LabelCollection(data=list(self))
+
+    def assign_schema_ids(
+            self, ontology_builder: OntologyBuilder) -> "LabelGenerator":
+
+        def _assign_ids(label: Label):
+            label.assign_schema_ids(ontology_builder)
+            return label
+
+        self._fns['assign_schema_ids'] = _assign_ids
+        return self
+
+    def add_url_to_data(self, signer: Callable[[bytes],
+                                               str]) -> "LabelGenerator":
+        """
+        Creates signed urls for the data
+        Only uploads url if one doesn't already exist.
+
+        Args:
+            signer: A function that accepts bytes and returns a signed url.
+        Returns:
+            LabelGenerator that signs urls as data is accessed
+        """
+
+        def _add_url_to_data(label: Label):
+            label.add_url_to_data(signer)
+            return label
+
+        self._fns['_add_url_to_data'] = _add_url_to_data
+        return self
+
+    def add_to_dataset(self, dataset: "Entity.Dataset",
+                       signer: Callable[[bytes], str]) -> "LabelGenerator":
+        """
+        Creates data rows from each labels data object and attaches the data to the given dataset.
+        Updates the label's data object to have the same external_id and uid as the data row.
+
+        This is a lot slower than LabelCollection.add_to_dataset but also more memory efficient.
+
+        Args:
+            dataset: labelbox dataset object to add the new data row to
+            signer: A function that accepts bytes and returns a signed url.
+        Returns:
+            LabelGenerator that updates references to the new data rows as data is accessed
+        """
+
+        def _add_to_dataset(label: Label):
+            label.create_data_row(dataset, signer)
```

Comment:
I was wondering how the generator version looks so much simpler than the collection version?

Reply:
It uses the functions provided by the Label class because it operates on each row individually. The Collection class tries to run bulk operations. They are still fairly similar in terms of lines of code.

Reply:
Got it, thanks